Add support for usernames in git resolver (Take 2)

danharper opened the pull request:
**Summary**

<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->

The Git resolver did not understand usernames in URLs, e.g. `git+ssh://git@github.com/...` due to it treating any pattern containing `@` as being "un-exploded". See https://github.com/yarnpkg/yarn/commit/411205015b7d4f4af3c344062c1f85c64a32365c, #513.

For me it was falling through, and calling the npm resolver instead. The package wasn't published on the registry, so it was erroring there. For others, it would error due to version mismatches if it was on the registry.

Instead, it's now simply treating any `pattern` with no protocol as being invalid. [There's no valid npm git URL without a protocol.](https://docs.npmjs.com/files/package.json#git-urls-as-dependencies)

Example failing pattern from before:

```
eslint-config-radweb@git+https://git@github.com/radweb/eslint-config-radweb.git
                    ^               ^
```

**Test plan**

<!-- Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots / videos if the pull request changes UI. -->

Added test, they pass.

Btw I've noticed `https://github.com/npm-ml/ocaml.git#npm-4.02.3` times out a _lot_.

kittens commented:
Can you please rebase, there's a lot of unnecessary files here

danharper commented:
Sure, I'll rebase in the morning. The additional files in the second commit
were added by the test runner. I was unsure if they should be committed?

On Fri, 14 Oct 2016, 22:28 Sebastian McKenzie, <notifications@github.com>
wrote:

> Can you please rebase, there's a lot of unnecessary files here
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/yarnpkg/yarn/pull/934#issuecomment-253923946>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAfLFELy3JB1Ws9uqIqqkR8CMtd-8Phxks5qz_PngaJpZM4KVPuG>
> .
>

danharper commented:
@kittens updated. The second commit still contains the additional test fixtures which Jest produced.

EugeneHlushko commented:
Hi @kittens, can we proceed with this PR? Looking forward to move yarn into few projects but git+ssh url from private git repo is still not found:
`error Couldn't find package "HIDDENTEXT" on the "npm" registry.`

danharper commented:
I've rebased the PR with master.

I've also removed the second commit containing all the `.bin` fixtures generated by Jest. Looking at #1093 it seems these shouldn't be generating, but they're created every time I run tests locally.

joezcool02 commented:
this is the fix I need for my project!

EugeneHlushko commented:
Hi @kittens , can you comment in why this is stalled? Maybe we can help out
